poptorch_experimental_addons/sharded.py

```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+from typing import Any
+
+import einops
+import torch
+
+from .collectives import all_gather_cross_replica, all_reduce_cross_replica_sum
+
+
+def rowcolrow_sharded_matmul(
+    X: torch.Tensor, Y: torch.Tensor, replication_factor: int, num_chunks: int = 1
+) -> Any:
+    """
+    Matrix multiplation for row-sharded x column-sharded -> row-sharded tensors
+
+    Gathers the right multiplicand across IPU program replicas
+    """
+    X_local_outer_dim, _ = X.shape
+    _, Y_local_outer_dim = Y.shape
+    result = torch.zeros((X_local_outer_dim, Y_local_outer_dim * replication_factor))
+    X = einops.rearrange(X, "m (k c) -> m k c", k=num_chunks)
+    Y = einops.rearrange(Y, "(k c) n -> c k n", k=num_chunks)
+    for i in range(num_chunks):
+        index = torch.tensor([i])
+        Yg = all_gather_cross_replica(
+            torch.index_select(Y, dim=1, index=index).squeeze(), replication_factor
```

Comment:
I thought you were having to use `index_select` for dynamic indices, but this is static. Out of curiosity, what happens if you do `Y[:, i]`?

Reply:
```
poptorch.poptorch_core.Error: In unknown:0: 'popart_exception': Failure in BasePadOp::runtimeConfirmShapes, for 390 (ai.graphcore.SliceGrad:1). This with pads of size 4, and input tensor of rank 1. Expected pads to be exactly 2x the unput Tensor's rank.
```

Reply:
Thanks for this—depressing, but good to know!